src/Spanner/ValueMapper.php

```diff
@@ -0,0 +1,328 @@
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner;
+
+use Google\Cloud\Core\ArrayTrait;
+use Google\Cloud\Core\Int64;
+use google\spanner\v1\TypeCode;
+
+/**
+ * Manage value mappings between Google Cloud PHP and Cloud Spanner
+ */
+class ValueMapper
+{
+    use ArrayTrait;
+
+    const NANO_REGEX = '/\.(\d{1,9})Z/';
+
+    const TYPE_BOOL = TypeCode::TYPE_BOOL;
+    const TYPE_INT64 = TypeCode::TYPE_INT64;
+    const TYPE_FLOAT64 = TypeCode::TYPE_FLOAT64;
+    const TYPE_TIMESTAMP = TypeCode::TYPE_TIMESTAMP;
+    const TYPE_DATE = TypeCode::TYPE_DATE;
+    const TYPE_STRING = TypeCode::TYPE_STRING;
+    const TYPE_BYTES = TypeCode::TYPE_BYTES;
+    const TYPE_ARRAY = TypeCode::TYPE_ARRAY;
+    const TYPE_STRUCT = TypeCode::TYPE_STRUCT;
+    const TYPE_NULL = 'null';
+
+    /**
+     * @var bool
+     */
+    private $returnInt64AsObject;
+
+    /**
+     * @param bool $returnInt64AsObject
+     */
+    public function __construct($returnInt64AsObject)
+    {
+        $this->returnInt64AsObject = $returnInt64AsObject;
+    }
+
+    /**
+     * Accepts an array of key/value pairs, where the key is a SQL parameter
+     * name and the value is the value interpolated by the server, and returns
+     * an array of parameters and inferred parameter types.
+     *
+     * @param array $parameters The key/value parameters.
+     * @return array An associative array containing params and paramTypes.
+     */
+    public function formatParamsForExecuteSql(array $parameters)
+    {
+        $paramTypes = [];
+
+        foreach ($parameters as $key => $value) {
+            list ($parameters[$key], $paramTypes[$key]) = $this->paramType($value);
+        }
+
+        return [
+            'params' => $parameters,
+            'paramTypes' => $paramTypes
+        ];
+    }
+
+    /**
+     * Accepts a list of values and encodes the value into a format accepted by
+     * the Spanner API.
+     *
+     * @param array $values The list of values
+     * @return array The encoded values
+     */
+    public function encodeValuesAsSimpleType(array $values)
+    {
+        $res = [];
+        foreach ($values as $value) {
+            $res[] = $this->paramType($value)[0];
+        }
+
+        return $res;
+    }
+
+    /**
+     * Accepts a list of columns (with name and type) and a row from read or
+     * executeSql and decodes each value to its corresponding PHP type.
+     *
+     * @param array $columns The list of columns.
+     * @param array $row The row data.
+     * @return array The decoded row data.
+     */
+    public function decodeValues(array $columns, array $row)
+    {
+        $cols = [];
+        $types = [];
+
+        foreach ($columns as $index => $column) {
+            $cols[] = (isset($column['name']))
+                ? $column['name']
+                : $index;
+            $types[] = $column['type'];
+        }
+
+        $res = [];
+        foreach ($row as $index => $value) {
+            $i = $cols[$index];
```

Comment:
We have a working solution for this case now. When calling `$results->rows()`  you can now specify how you would like your results to be formatted.

```php
    /**
     * Return the formatted and decoded rows. If the stream is interrupted and
     * a resume token is available, attempts will be made on your behalf to
     * resume.
     *
     * Example:
     * ```
     * $rows = $result->rows();
     * ```
     *
     * @param string @format Determines the format in which rows are returned.
     *        `Result::RETURN_NAME_VALUE_PAIR` returns items as a
     *        multi-dimensional array containing a name and a value key.
     *        Ex: `[0 => ['name' => 'column1', 'value' => 'my_value']]`.
     *        `Result::RETURN_ASSOCIATIVE` returns items as an associative array
     *        with the column name as the key. Please note with this option, if
     *        duplicate columns are present the value from the latter column
     *        will be retained. Ex: `['column1' => 'my_value']`.
     *        `Result::RETURN_ZERO_INDEXED` returns items as a 0 indexed array,
     *        with the key representing the column number as found by executing
     *        {@see Google\Cloud\Spanner\Result::columns()}.
     *        Ex: `[0 => 'my_value']`. **Defaults to**
     *        `Result::RETURN_NAME_VALUE_PAIR`.
     * @return \Generator
     */
    public function rows($format = self::RETURN_NAME_VALUE_PAIR)
```

I am just combing through and updating tests/documentation to reflect this change now and will have a commit ready soon.